lib/ask/survey_respondents_canceller.ex

```diff
@@ -1,111 +1,65 @@
-        :ignore
-
-      _ ->
-        state = %{survey_ids: survey_ids, last_updated_respondent_id: 0}
-        {:producer, state}
-    end
-  end
+  alias Ecto.Multi
 
-  def handle_demand(_demand, state) do
-    respondent_ids = get_respondents_for_update(state)
-    handle_respondents(respondent_ids, state)
+  def start_link(survey_id) do
+    GenServer.start_link(__MODULE__, survey_id)
   end
 
-  def handle_respondents(respondent_ids, state) when length(respondent_ids) > 0 do
-    last_id = List.last(respondent_ids)
-    new_state = %{survey_ids: state.survey_ids, last_updated_respondent_id: last_id}
-    {:noreply, respondent_ids, new_state}
+  def cancel(survey_id) do
+    SurveyCancellerSupervisor.start_cancelling(survey_id)
   end
 
-  def handle_respondents(respondent_ids, state) when length(respondent_ids) == 0 do
-    state.survey_ids
-    |> Enum.each(fn survey_id ->
-      survey = Repo.get(Survey, survey_id)
-      project = Repo.get!(Project, survey.project_id)
-
-      changeset =
-        Survey.changeset(survey, %{
-          state: "terminated",
-          exit_code: 1,
-          exit_message: "Cancelled by user"
-        })
-
-      Multi.new()
-      |> Multi.update(:survey, changeset)
-      |> Multi.insert(:log, ActivityLog.completed_cancel(project, nil, survey))
-      |> Repo.transaction()
-    end)
-
-    {:stop, :normal, state}
+  @impl true
+  def init(survey_id) do
+    :timer.send_after(1000, :cancel)
+    Logger.info("Starting canceller for survey #{survey_id}")
+    {:ok, survey_id}
   end
 
-  def get_respondents_for_update(state) do
-    query =
+  defp respondents_to_cancel(survey_id) do
+    Repo.all(
       from(
         r in Respondent,
         select: r.id,
-        where:
-          r.state == :active and r.survey_id in ^state.survey_ids and
-            r.id > ^state.last_updated_respondent_id,
-        limit: 100,
-        order_by: [
-          asc: r.id
-        ]
+        where: r.state == :active and r.survey_id == ^survey_id,
+        limit: 100
       )
-
-    Repo.all(query)
-  end
-end
-
-defmodule Ask.RespondentsCancellerConsumer do
-  use Ecto.Schema
-  alias Ask.{RespondentsCancellerProducer, Respondent, Repo}
-  alias Ask.Runtime.Session
-  use GenStage, restart: :transient
-
-  def start_link(_initial) do
-    GenStage.start_link(__MODULE__, :state, name: __MODULE__)
+    )
   end
 
-  def init(producer_pid) do
-    {:consumer, :state, subscribe_to: [{producer_pid, max_demand: 50, min_demand: 1}]}
-  end
-
-  def handle_events(respondent_ids, _from, state) do
-    respondent_ids
-    |> Enum.each(fn respondent_id -> Respondent.with_lock(respondent_id, &cancel_respondent/1) end)
-
-    {:noreply, [], state}
+  defp cancel_survey(survey_id) do
+    survey = Repo.get(Survey, survey_id)
+    project = Repo.get!(Project, survey.project_id)
+
+    changeset =
+      Survey.changeset(survey, %{
+        state: "terminated",
+        exit_code: 1,
+        exit_message: "Cancelled by user"
```

Comment:
at this point we don't know if it was cancelled by user or not. I think we should't change the exit message at this point

Reply:
I don't think there's any other way in which a survey gets cancelled - only if a user presses the Stop button.

If a survey hits its target, it'd exhaust the then-active respondents, then finish. If a survey exhausts all its respondents, it'll just finish. But I think there's no other way to cancel a survey - am I missing something?

Reply:
The `survey.exit_message` is already being modified/defined in `SurveyAction` when we change the `survey.state` to `cancelling`. So, adding this to the changeset at this level doesn't really add value to the survey behavior.

So the comment is mostly "technical". It doesn't add value to the final behavior and IMHO worsens a bit the codebase maintainability. If in the future we decide to add another way of cancelling a Survey, we should modify this component as well when this component actually doesn't really care _why_ it was cancelled

Reply:
Oh, great observation. I've also dropped the `exit_code: 1` part, too, since it's also set by the same `SurveyAction`.